File: 238.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> output(n, 1);

        // First pass: calculate the prefix products
        int prefix_product = 1;
        for (int i = 0; i < n; ++i) {
            output[i] = prefix_product;
            prefix_product *= nums[i];
        }

        // Second pass: calculate the suffix products and update the output
        int suffix_product = 1;
        for (int i = n - 1; i >= 0; --i) {
            output[i] *= suffix_product;
            suffix_product *= nums[i];
        }

        return output;
    }
};

// Time Complexity: O(n) — We go through the array twice (once for the left products and once for the right products), so the time complexity is linear in terms of the array length.
// Space Complexity: O(1) (ignoring the output array) — We only use a few extra variables (left_product and right_product), so the space complexity is constant (apart from the output array).

File: 238.py
def productExceptSelf(nums):
    n = len(nums)
    output = [1] * n  # Step 1: Initialize the output array with 1

    # Step 2: Calculate the product of all elements to the left of each index
    left_product = 1
    for i in range(n):
        output[i] = left_product
        left_product *= nums[i]

    # Step 3: Calculate the product of all elements to the right of each index
    right_product = 1
    for i in range(n - 1, -1, -1):
        output[i] *= right_product
        right_product *= nums[i]

    return output